pkg/controller/cronjob/cronjob_controller2.go

```diff
@@ -0,0 +1,580 @@
+		// No controller should care about orphans being deleted.
+		return
+	}
+	cronJob := jm.resolveControllerRef(job.Namespace, controllerRef)
+	if cronJob == nil {
+		return
+	}
+	jm.enqueueController(cronJob)
+}
+
+func (jm *Controller2) enqueueController(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+
+	jm.queue.Add(key)
+}
+
+func (jm *Controller2) enqueueControllerWithTime(obj interface{}, t time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+
+	jm.queue.AddAfter(key, t)
+}
+
+func (jm *Controller2) updateCronJob(old interface{}, curr interface{}) {
+	//  the big cases that needs to be handled
+	//  1. change in CJ.spec.schedule
+	//  2. change in CJ.spec.JobTemplate
+	//  other CJ.spec params seem to be dealt with by the controller in sync loop on updates
+	oldCJ, okOld := old.(*batchv1beta1.CronJob)
+	newCJ, okNew := curr.(*batchv1beta1.CronJob)
+
+	if !okOld || !okNew {
+		// typecasting of one failed, handle this better, may be log entry
+		return
+	}
+	// if the change in schedule results in next requeue having to be sooner than it already was,
+	// it will be handled here by the queue. If the next requeue is further than previous schedule,
+	// the sync loop will essentially be a no-op for the already queued key with old schedule.
+	if oldCJ.Spec.Schedule != newCJ.Spec.Schedule {
+		// schedule changed, change the requeue time
+		sched, err := cron.ParseStandard(newCJ.Spec.Schedule)
+		if err != nil {
+			// this is likely a user error in defining the spec value
+			// we should log the error and not reconcile this cronjob until an update to spec
+			klog.Errorf("unparseable schedule: %s : %s", newCJ.Spec.Schedule, err)
+			return
+		}
+		now := jm.now()
+		t := sched.Next(now).Sub(now)
+
+		jm.enqueueControllerWithTime(curr, t)
+		return
+	}
+
+	// other parameters changed, requeue this now and if this gets triggered
+	// within deadline, sync loop will work on the CJ otherwise updates will be handled
+	// during the next schedule
+	jm.enqueueController(curr)
+}
+
+// TODO: @alpatel we need to return errors from this function, in order to allow
+//  for errors to propagate up the reconcile function and kick in queues exponential
+//  backed off retries more details on inline code comments in the func
+
+// syncOne reconciles a CronJob with a list of any Jobs that it created.
+// All known jobs created by "cj" should be included in "js".
+// The current time is passed in to facilitate testing.
+// It has no receiver, to facilitate testing.
+func syncOne2(
+	cj *batchv1beta1.CronJob,
+	js []batchv1.Job,
+	now time.Time,
+	jc jobControlInterface,
+	cjc cjControlInterface,
+	recorder record.EventRecorder) (error, *time.Duration) {
+	nameForLog := fmt.Sprintf("%s/%s", cj.Namespace, cj.Name)
```

Comment:
Like mentioned before use strucutured logging for this.

Reply:
updated to use structured logging